### PR TITLE
Batch by dimension in Sfx handler

### PR DESCRIPTION
**Pull request by venkey-ariv**
This PR adds the ability to,

1. Batch metrics by a dimension name at the handler level. Currently implemented for Sfx handler
2. For each batch, define it's own auth token and use the same when emitting metrics to the providers endpoint

### CODE REVIEW

**Comment by poros** [comment]
I also wanted to add @kokje to the reviewers list, but the interface doesn't seem to like the username...

File: src/fullerite/handler/handler.go [hunk 1]
```diff
@@ -109,6 +109,11 @@ type Handler interface {
 	SetCollectorWhiteList([]string)
 	CollectorWhiteList() map[string]bool
 	IsCollectorWhiteListed(string) (bool, bool)
+
+	// Return true if handler implementation
+	// takes care of reporting emission metrics
+	OverrideBaseEmissionMetricsReporter()
```
Comment (kokje) [inline src/fullerite/handler/handler.go]: Would be nice if this follows the same naming convention set and is

File: src/fullerite/handler/handler.go [hunk 2]
```diff
@@ -509,34 +537,33 @@ func (base *BaseHandler) recordEmissions(timingsChannel <-chan emissionTiming) {
...
-	metrics []metric.Metric,
-	emitFunc func([]metric.Metric) bool,
-	callbackChannel chan<- emissionTiming,
-) {
-	numMetrics := len(metrics)
-	beforeEmission := time.Now()
-	result := emitFunc(metrics)
-	afterEmission := time.Now()
+func (base *BaseHandler) reportEmissionMetrics(emissionResult bool, timing emissionTiming) {
+	base.log.Info(
```
Comment (kokje) [inline src/fullerite/handler/handler.go]: We should log only is emissionResult is true?

File: src/fullerite/handler/signalfx.go [hunk 3]
```diff
@@ -186,10 +242,35 @@ func (s *SignalFx) emitMetrics(metrics []metric.Metric) bool {
...
+	}
+	return emissionResult, elapsed
 }
 
-func signalFxKeySanitize(key string) string {
-	return util.StrSanitize(key, false, allowedDimKeyPuncts)
-}
+func (s *SignalFx) emitMetrics(metrics []metric.Metric) bool {
+
+	if len(metrics) == 0 {
```
Comment (kokje) [inline src/fullerite/handler/signalfx.go]: Is this duplicated on line 189 too?

File: src/fullerite/handler/signalfx.go [hunk 4]
```diff
@@ -130,7 +164,26 @@ func (s SignalFx) getSanitizedDimensions(incomingMetric metric.Metric) map[strin
...
+		return authToken
+	} else {
+		return s.authToken
+	}
+}
+
+func (s *SignalFx) makeBatches(metrics []metric.Metric) map[string][]metric.Metric {
+	m := make(map[string][]metric.Metric)
+	for _, metric := range metrics {
+		dimValue := metric.Dimensions[s.batchByDimension]
```
Comment (kokje) [inline src/fullerite/handler/signalfx.go]: What if this doesn't exist?

File: src/fullerite/handler/signalfx.go [hunk 5]
```diff
@@ -186,10 +242,35 @@ func (s *SignalFx) emitMetrics(metrics []metric.Metric) bool {
...
+	// Report emission metrics if emission tracker is disabled in base handler
+	if s.UseCustomEmissionMetricsReporter() {
+		timing := emissionTiming{
+				timestamp:   time.Now(),
+				duration:    elapsed,
+				metricsSent: len(metrics),
+			}
+		s.reportEmissionMetrics(emissionResult, timing)
+	}
+	return emissionResult, elapsed
```
Comment (kokje) [inline src/fullerite/handler/signalfx.go]: Why do we need to return something here?

File: src/fullerite/handler/handler.go (hunk 1 above)
Comment (venkey-ariv) [inline src/fullerite/handler/handler.go]: The set/is convention doesn't fit nicely here :( So I left it be

**Comment by venkey-ariv** [comment]
Discarded in favor of https://github.com/Yelp/fullerite/pull/327

**Pull request closed without merging**